### ISSUE TITLE
Cannot enable Datadog Appsec - libddwaf failed to load

### ISSUE THREAD

**Issue by ricksuggs** [DataDog/dd-trace-rb#2652]
**Current behaviour**
I'm trying to enable Datadog appsec on an application, but libddawf cannot be found where it should have been installed.

**Expected behaviour**
libddawf should be found

**Steps to reproduce**
I found the following log entries that suggest a dependency cannot be loaded:

```
E, [2023-02-28T20:21:30.953833 #11] ERROR -- ddtrace: [ddtrace] (/app/vendor/bundle/ruby/3.0.0/gems/ddtrace-1.9.0/lib/datadog/appsec/processor.rb:172:in `rescue in require_libddwaf') libddwaf failed to load,installed platform: ruby ruby platforms: ["ruby", "x86_64-linux"] error: #<LoadError: Could not open library '/app/vendor/bundle/ruby/3.0.0/gems/libddwaf-1.5.1.0.1/lib/datadog/appsec/../../../vendor/libddwaf/libddwaf-1.5.1-linux-gnu-x86_64/lib/libddwaf.so': /app/vendor/bundle/ruby/3.0.0/gems/libddwaf-1.5.1.0.1/lib/datadog/appsec/../../../vendor/libddwaf/libddwaf-1.5.1-linux-gnu-x86_64/lib/libddwaf.so: cannot open shared object file: No such file or directory>

W, [2023-02-28T20:21:30.953881 #11]  WARN -- ddtrace: [ddtrace] AppSec is disabled, see logged errors above
```

Dependency versions:

ddtrace ruby gem version: 1.9.0
datadog-agent version: 7.43.0
datadog-heroku-buildpack version: 2.9
ruby version: 3.0.5
rails version: 6.1.7.1

Output of agent-wrapper status command:

```
~ $ agent-wrapper status
Getting the status from the agent.

===============
Agent (v7.43.0)
===============

  Status date: 2023-02-28 17:51:58.217 UTC (1677606718217)
  Agent start: 2023-02-28 17:09:33.015 UTC (1677604173015)
  Pid: 81
  Go Version: go1.19.5
  Python Version: 3.8.16
  Build arch: amd64
  Agent flavor: agent
  Check Runners: 4
  Log File: /app/.apt/var/log/datadog/datadog.log
  Log Level: DEBUG

  Paths
  =====
    Config File: /app/.apt/etc/datadog-agent/datadog.yaml
    conf.d: /app/.apt/etc/datadog-agent/conf.d
    checks.d: /app/.apt/etc/datadog-agent/checks.d

  Clocks
  ======
    NTP offset: 43µs
    System time: 2023-02-28 17:51:58.217 UTC (1677606718217)

  Host Info
  =========
    bootTime: 2023-02-10 00:14:02 UTC (1675988042000)
    hostId: redacted
    kernelArch: x86_64
    kernelVersion: 4.4.0-1104-aws
    os: linux
    platform: ubuntu
    platformFamily: debian
    platformVersion: 20.04
    procs: 7
    uptime: 448h55m33s
```

**Comment by GustavoCaso** [comment]
Hi @ricksuggs 

By looking at the logline: 
```
libddwaf failed to load,installed platform: ruby ruby platforms: ["ruby", "x86_64-linux"]
```

The `installed platform: ruby` section indicates that the libddwaf gem has been fetched and installed as ruby. 
It is important to note that the ruby gem platform means “build from source”, which as of today the libddwaf gem is unable to do.

However, we want it to install the `x86_64-linux` platform instead, since ASM is supported on that platform. 

Could you confirm any of these points: 

- Are you using the  --force-ruby-platform flag on the bundler command line?
- Can you confirm the version of bundler and/or rubygems versions?
- Is `x86_64-linux` included in the PLATFORMS section of Gemfile.lock? When the lock file is being generated on another platform, this may be resolved by using `bundle lock --add-platform x86_64-linux`

**Comment by ricksuggs** [comment]
I'm not using the --force-ruby-platform flag.
I'm using the heroku ruby buildpack, which always uses the latest stable version of bundler and rubygems.
That platform was not listed in the platforms section of the Gemfile.lock, I will try adding it.
Thank you

**Comment by ricksuggs** [comment]
@GustavoCaso Using `bundle lock --add-platform x86_64-linux`, I added the platform to the Gemfile.lock. This fixed the issue with DataDog appsec, but now datadog profiling is disabled because it can't find a different dependency for that platform.

```
W, [2023-03-01T16:15:04.448688 #13]  WARN -- ddtrace: [ddtrace] Profiling was requested but is not supported, profiling disabled: There was an error loading the profiling native extension due to 'RuntimeError Failure to load ddtrace_profiling_native_extension.3.0.5_x86_64-linux due to libdatadog_profiling.so: cannot open shared object file: No such file or directory' at '/app/vendor/bundle/ruby/3.0.0/gems/ddtrace-1.9.0/lib/datadog/profiling/load_native_extension.rb:22:in `<top (required)>''
```

**Comment by ivoanjo** [comment]
Hey there @ricksuggs :wave: I'm Ivo and I work with Gustavo (I focus on the profiling bits!).

Sorry that you run into this second issue. 

What I suspect happened is that the `bundle lock --add-platform x86_64-linux` command also changed the version of the `libdatadog` gem from `1.0.1.1.0` to `1.0.1.1.0-x86_64-linux`. (`libdatadog` is a dependency of profiling).

This should not normally be a problem but due to a few limitations in rubygems that we haven't found a solution for (#2242), the `ddtrace` gem doesn't get recompiled when only the version of `libdatadog` changes, and so it tries to use the old version (which in your case isn't there anymore).

TL;DR: Purging the heroku build cache should fix the issue, because it will cause `ddtrace` to be installed from scratch, with the correct version of the `libdatadog` gem.

See https://help.heroku.com/18PI5RSY/how-do-i-clear-the-build-cache and https://github.com/heroku/heroku-builds for instructions on how to clear the build cache.

**Comment by ricksuggs** [comment]
@ivoanjo Clearing the build cache worked. Profiles and Appsec are both working now. Thank you!

**Comment by nertzy** [comment]
Related: https://github.com/DataDog/libddwaf-rb/issues/32
